File: src/main/java/com/cuupa/mailprocessor/process/AbstractProcessInstanceHandler.java
package com.cuupa.mailprocessor.process;

import org.camunda.bpm.engine.delegate.DelegateExecution;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public abstract class AbstractProcessInstanceHandler {

    private final DelegateExecution delegateExecution;

    public AbstractProcessInstanceHandler(final DelegateExecution delegateExecution) {
        this.delegateExecution = delegateExecution;
        for(ProcessProperty property : ProcessProperty.values()) {
            if(!delegateExecution.hasVariable(property.name())){
                delegateExecution.setVariable(property.name(), null);
            }
        }
    }

    protected void add(final String key, final Object value) {
        if (delegateExecution.hasVariable(key)) {
            final Map<String, Object> variableCasted = getAsT(key, Map.class);
            delegateExecution.setVariable(key, variableCasted);
        } else {
            delegateExecution.setVariable(key, new ArrayList<>());
        }
    }

    protected void set(final String key, final Object value) {
        delegateExecution.setVariable(key, value);
    }

    protected String getAsString(final String propertyName) {
        if(delegateExecution.hasVariable(propertyName)){
            return getAsT(propertyName, String.class);
        }
        return null;
    }

    protected <T> T getAsListOf(final String propertyName){
        return getAsT(propertyName, List.class);
    }

    protected List<String> getAsListOfString(final String propertyName) {
        if (delegateExecution.hasVariable(propertyName)) {
            return getAsT(propertyName, List.class);
        }
        return new ArrayList<>();
    }

    @NotNull
    private <T> T getAsT(final String propertyName, Class clazz) {
        final Object property = delegateExecution.getVariable(propertyName);
        if (clazz.isInstance(property)) {
            return (T) clazz.cast(property);
        } else {
            throw new RuntimeException("Invalid data type");
        }
    }

    protected Map<String, Object> getAsMap(final String propertyName) {
        if(delegateExecution.hasVariable(propertyName)){
            return getAsT(propertyName, Map.class);
        }
        return new HashMap<>();
    }

    protected byte[] getAsByteArray(final String propertyName) {
        if (delegateExecution.hasVariable(propertyName)) {
            return getAsT(propertyName, byte[].class);
        }
        return new byte[0];
    }

    protected DelegateExecution getDelegateExecution(){
        return delegateExecution;
    }
}

File: src/main/java/com/cuupa/mailprocessor/process/AbstractProcessInstanceHandler.kt
package com.cuupa.mailprocessor.process

import org.camunda.bpm.engine.delegate.DelegateExecution
import java.time.LocalDateTime
import java.util.*
import kotlin.collections.ArrayList

abstract class AbstractProcessInstanceHandler(protected val delegateExecution: DelegateExecution) {
    //FIXME: this can't work
    protected fun add(property: ProcessProperty, value: Any?) {
        if (delegateExecution.hasVariable(property.name) && delegateExecution.getVariable(property.name) != null) {
            val variableCasted = getAsT<MutableMap<String, Any?>>(property)
            variableCasted[property.name] = value
            delegateExecution.setVariable(property.name, variableCasted)
        } else {
            delegateExecution.setVariable(property.name, listOf(value))
        }
    }

    protected fun addToList(property: ProcessProperty, value: Any?) {
        if (delegateExecution.hasVariable(property.name) && delegateExecution.getVariable(property.name) != null) {
            val variableCasted = getAsT<MutableList<Any?>>(property)
            delegateExecution.setVariable(property.name, addIfApplicable(value, variableCasted))
        } else {
            val list = mutableListOf<Any?>()
            delegateExecution.setVariable(property.name, addIfApplicable(value, list))
        }
    }

    private fun addIfApplicable(value: Any?, list: MutableList<Any?>): MutableList<Any?> {
        if (value is Collection<*>) {
            if (value.isEmpty()) {
                return list
            }
            list.addAll(value)
        } else {
            list.add(value)
        }
        return list
    }

    protected operator fun set(property: ProcessProperty, value: Any?) {
        delegateExecution.setVariable(property.name, value)
    }

    protected fun getAsString(propertyName: ProcessProperty): String? {
        return if (delegateExecution.hasVariable(propertyName.name)) {
            getAsT<String>(propertyName)
        } else null
    }

    protected fun getAsLocalDateTime(propertyName: ProcessProperty): LocalDateTime? {
        return if (delegateExecution.hasVariable(propertyName.name)) {
            getAsT<LocalDateTime>(propertyName)
        } else null
    }

    protected fun <T> getAsListOf(propertyName: ProcessProperty): List<T> {
        return getAsT<MutableList<T>?>(propertyName) ?: mutableListOf()
    }

    protected fun getAsListOfString(propertyName: ProcessProperty): List<String> {
        return if (delegateExecution.hasVariable(propertyName.name)) {
            getAsT(propertyName) ?: ArrayList()
        } else ArrayList()
    }

    private fun <T> getAsT(propertyName: ProcessProperty): T {
        return delegateExecution.getVariable(propertyName.name) as T
    }

    protected fun getAsMap(propertyName: ProcessProperty): Map<String, Any> {
        return if (delegateExecution.hasVariable(propertyName.name)) {
            getAsT(propertyName) ?: mapOf()
        } else HashMap()
    }

    protected fun getAsByteArray(propertyName: ProcessProperty): ByteArray? {
        return if (delegateExecution.hasVariable(propertyName.name)) {
            getAsT(propertyName)
        } else ByteArray(0)
    }

    protected fun getAsBooleanDefaultFalse(propertyName: ProcessProperty): Boolean {
        return if (delegateExecution.hasVariable(propertyName.name)) {
            return delegateExecution.getVariable(propertyName.name) as Boolean? ?: false
        } else {
            false
        }
    }

    init {
        for (property in ProcessProperty.values()) {
            if (!delegateExecution.hasVariable(property.name)) {
                delegateExecution.setVariable(property.name, null)
            }
        }
    }
}